providers/asrockrack/asrockrack.go

```diff
@@ -0,0 +1,225 @@
+	if a.fwInfo == nil {
+		a.fwInfo, err = a.firmwareInfo()
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return a.fwInfo.BIOSVersion, nil
+}
+
+// nolint: gocyclo
+// BMC firmware update is a multi step process
+// this method initiates the upgrade process and waits in a loop until the device has been upgraded
+func (a *ASRockRack) FirmwareUpdateBMC(ctx context.Context, filePath string) error {
+
+	defer func() {
+		// The device needs to be reset to be removed from flash mode,
+		// this is required once setFlashMode() is invoked.
+		// The BMC resets itself once a firmware flash is successful or failed.
+		if a.resetRequired {
+			a.log.V(1).Info("info", "resetting BMC, this takes a few minutes")
+			err := a.reset()
+			if err != nil {
+				a.log.Error(err, "failed to reset BMC")
+			}
+		}
+	}()
+
+	// 0. validate the given filePath exists
+	_, err := os.Stat(filePath)
+	if os.IsNotExist(err) {
+		return err
+	}
+
+	// 1. set the device to flash mode - prepares the flash
+	a.log.V(0).Info("info", "step", "1/5 - setting device into flash mode.. this takes a minute")
+	err = a.setFlashMode()
+	if err != nil {
+		return fmt.Errorf("failed in step 1/5 - set device in flash mode: " + err.Error())
+	}
+
+	// 2. upload firmware image file
+	a.log.V(0).Info("info", "step", "2/5 uploading firmware image", "filePath", filePath)
+	err = a.uploadFirmware(filePath)
+	if err != nil {
+		return fmt.Errorf("failed in step 2/5 - upload firmware image: " + err.Error())
+	}
+
+	// 3. BMC to verify the uploaded file
+	err = a.verifyUploadedFirmware()
+	a.log.V(0).Info("info", "step", "3/5 - bmc to verify uploaded firmware")
+	if err != nil {
+		return fmt.Errorf("failed in step 3/5 - verify uploaded firmware: " + err.Error())
+	}
+
+	startTS := time.Now()
+	// 4. Run the upgrade - preserving current config
+	a.log.V(0).Info("info", "step 4/5", "run the upgrade, preserving current configuration")
+	err = a.upgradeBMC()
+	if err != nil {
+		return fmt.Errorf("failed in step 4/5 - verify uploaded firmware: " + err.Error())
+	}
+
+	// progress check interval
+	progressT := time.NewTicker(2 * time.Second).C
+	// timeout interval
+	timeoutT := time.NewTicker(30 * time.Minute).C
+	maxErrors := 20
+	var errorsCount int
+
+	// 5.loop until firmware was updated - with a timeout
+	for {
+		select {
+		case <-progressT:
+			// check progress
+			p, err := a.flashProgress()
+			if err != nil {
+				errorsCount++
+				a.log.V(0).Error(err, "step", "5/5 - error checking flash progress", "error count", errorsCount, "max errors", maxErrors, "elapsed time", time.Since(startTS).String())
+				continue
+			}
+
+			a.log.V(0).Info("info", "step", "5/5 - check flash progress..", "progress", p.Progress, "action", p.Action, "elapsed time", time.Since(startTS).String())
+
+			// all done!
+			if p.State == 2 {
+				a.log.V(0).Info("info", "step", "5/5 - firmware flash complete!", "progress", p.Progress, "action", p.Action, "elapsed time", time.Since(startTS).String())
```

Comment:
here's no way to turn of a log entry with `V(0).Info`. All these messages make for quite a chatty function. what do you think about making these messages opt-in through the `V(1).Info` level?